java-checks/src/main/java/org/sonar/java/checks/spring/NonSingletonAutowiredInSingletonCheck.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.spring;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6832")
+public class NonSingletonAutowiredInSingletonCheck extends IssuableSubscriptionVisitor {
+  private static final String SCOPED_ANNOTATION = "org.springframework.context.annotation.Scope";
+  private static final String AUTOWIRED_ANNOTATION = "org.springframework.beans.factory.annotation.Autowired";
+  private static final String JAVAX_INJECT_ANNOTATION = "javax.inject.Inject";
+  private static final String JAKARTA_INJECT_ANNOTATION = "jakarta.inject.Inject";
+  private static final Set<String> AUTO_WIRING_ANNOTATIONS = Set.of(AUTOWIRED_ANNOTATION, JAVAX_INJECT_ANNOTATION, JAKARTA_INJECT_ANNOTATION);
+  private static final Set<String> SINGLETON_LITERALS = Set.of("singleton", "\"singleton\"");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.ANNOTATION, Tree.Kind.CONSTRUCTOR);
+  }
+
+  /**
+   * This rule reports an issue when a Singleton Bean auto-wires a non-Singleton Bean via constructor, field or method parameter.
+   */
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.ANNOTATION)) {
+      var annotationTree = (AnnotationTree) tree;
+
+      if (isAutoWiringAnnotation(annotationTree) && isAnnotationOnKind(annotationTree, Tree.Kind.VARIABLE)) {
```

Comment:
I think we can make the conditions here a little easier to understand. You could extract `isAutoWiringAnnotation` to earlier in the check, maybe just return if it is false.
Instead of calling `isAnnotationOnKind` 3 times, you make the method return the annotated node instead, allowing you to reduce the conditions here to `annotatedNode.is(...)`. Another advantage of this is you don't have to get the annotation symbol again inside of the logic.

Reply:
let's discuss how to improve this